Content.Shared/Storage/EntitySystems/BinSystem.cs

```diff
@@ -0,0 +1,162 @@
+    [Dependency] private readonly SharedContainerSystem _container = default!;
+    [Dependency] private readonly SharedHandsSystem _hands = default!;
+
+    public const string BinContainerId = "bin-container";
+
+    /// <inheritdoc/>
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<BinComponent, ComponentGetState>(OnGetState);
+        SubscribeLocalEvent<BinComponent, ComponentHandleState>(OnHandleState);
+        SubscribeLocalEvent<BinComponent, ComponentStartup>(OnStartup);
+        SubscribeLocalEvent<BinComponent, MapInitEvent>(OnMapInit);
+        SubscribeLocalEvent<BinComponent, EntRemovedFromContainerMessage>(OnEntRemoved);
+        SubscribeLocalEvent<BinComponent, InteractHandEvent>(OnInteractHand);
+        SubscribeLocalEvent<BinComponent, AfterInteractUsingEvent>(OnAfterInteractUsing);
+    }
+
+    private void OnGetState(EntityUid uid, BinComponent component, ref ComponentGetState args)
+    {
+        args.State = new BinComponentState(component.Items, component.Whitelist, component.MaxItems);
+    }
+
+    private void OnHandleState(EntityUid uid, BinComponent component, ref ComponentHandleState args)
+    {
+        if (args.Current is not BinComponentState state)
+            return;
+
+        component.Items = new List<EntityUid>(state.Items);
+        component.Whitelist = state.Whitelist;
+        component.MaxItems = state.MaxItems;
+    }
+
+    private void OnStartup(EntityUid uid, BinComponent component, ComponentStartup args)
+    {
+        component.ItemContainer = _container.EnsureContainer<Container>(uid, BinContainerId);
+    }
+
+    private void OnMapInit(EntityUid uid, BinComponent component, MapInitEvent args)
+    {
+        // don't spawn on the client.
+        if (_net.IsClient)
+            return;
+
+        var xform = Transform(uid);
+        foreach (var id in component.InitialContents)
+        {
+            var ent = Spawn(id, xform.Coordinates);
+            if (!TryInsertIntoBin(uid, ent, component))
+            {
+                Logger.Error($"Entity {ToPrettyString(ent)} was unable to be initialized into bin {ToPrettyString(uid)}");
+                return;
+            }
+        }
+    }
+
+    private void OnEntRemoved(EntityUid uid, BinComponent component, EntRemovedFromContainerMessage args)
+    {
+        component.Items.Remove(args.Entity);
+    }
+
+    private void OnInteractHand(EntityUid uid, BinComponent component, InteractHandEvent args)
+    {
+        if (args.Handled || !_timing.IsFirstTimePredicted)
+            return;
+
+        EntityUid? toGrab = component.Items.LastOrDefault();
+        if (!TryRemoveFromBin(uid, toGrab, component))
+            return;
+
+        _hands.TryPickupAnyHand(args.User, toGrab.Value);
+        _admin.Add(LogType.Pickup, LogImpact.Low,
+            $"{ToPrettyString(uid):player} removed {ToPrettyString(toGrab.Value)} from bin {ToPrettyString(uid)}.");
+        args.Handled = true;
```

Comment:
Then you can make this (i personally like the hand pickup being inside of the method tho its up to you if  you want to leave it on the handler)
```suggestion
        if (!TryRemoveFromBin(uid, component: component))
            return;

        _admin.Add(LogType.Pickup, LogImpact.Low,
            $"{ToPrettyString(uid):player} removed {ToPrettyString(toGrab.Value)} from bin {ToPrettyString(uid)}.");
        args.Handled = true;
```